packages/polytechcolloscopeclient/lib/src/polytechcolloscopeclient_base.dart

```diff
@@ -0,0 +1,110 @@
+import 'package:beautiful_soup_dart/beautiful_soup.dart';
+import 'package:html/dom.dart';
+import 'package:requests_plus/requests_plus.dart';
+
+import 'colloscope_datastructs.dart';
+import 'consts.dart';
+
+class PolytechColloscopeClient {
+  final String username, password;
+
+  PolytechColloscopeClient(this.username, this.password) {
+    assert(username.isNotEmpty);
+    assert(password.isNotEmpty);
+  }
+
+  Future<List<Student>> fetchStudents(Year year) async {
+    var page = await RequestsPlus.get(Consts.kholleURL[year]!,
+        userName: username, password: password);
+
+    BeautifulSoup bs = BeautifulSoup(page.body);
+
+    //This isolates the student list part of the page, it consists in a list of <li> elements like
+    // <li><a href="?page=colles_2A&amp;id_etudiant=STUDENT_ID">INITAL. SUNRNAME</a> (groupe GROUPNUMBER)</li>
+    var studentsList = bs.find("table.colles")?.find("td > ul");
+
+    List<Student> students = [];
+
+    studentsList?.findAll("li > a").forEach((e) {
+      var name = e.innerHtml;
+      var link = e.getAttrValue("href");
+      var id = RegExp(r"id_etudiant=(\d+)").firstMatch(link!)?.group(1);
+
+      students.add(Student(name, id!));
+    });
+
+    return students;
+  }
+
+  Future<StudentColloscope> getColloscope(Year year, int studentId) async {
+    // TODO : Add a check to verify the studentid (look 5 lines below)
+
+    var page = await RequestsPlus.get(
+        Consts.khollesStudentURL[year]!
+            .replaceFirst(":id", studentId.toString()),
```

Comment:
Using a search and replace is like mistaking a bazooka for a fly. All you'd have to do is not put an id in khollesStudentURL and paste the id directly with a concatenation.
`"${Consts.khollesStudentURL[year]!}$studentId"`